default/python/universe_generation/galaxy.py

```diff
@@ -1,35 +1,349 @@
+        def parents(p1, children):
+            # print "pp p1 ", p1.pos, " children a ", children
+            if p1.parent:
+                # print "pp deeper ", p1.pos, " chlidren b ", children
+                children.append(p1)
+                return parents(p1.parent, children)
+            else:
+                # print "pp done p1 ", p1.pos, " children c ", children
+                return (p1, children)
+
+        if pos not in self.dsets:
+            return []
+
+        (root, children) = parents(self.dsets[pos], [])
+        # squash the chain of children
+        for child in children:
+            child.bind_parent(root)
+        return [root.pos]
+
+    def complete_sets(self):
+        """ returns a list of list of all sets O(n)."""
+        ret = defaultdict(list)
+        for pos in self.dsets.keys():
+            ret[self.root(pos)].append(pos)
+        return ret.values()
+
+
+class Clusterer(object):
+    """
+    Computes all clusters of positions separated by more than MAX_STARLANE_LENGTH
+
+    This creates a disjoint set of all of the positions
+    Adds all of the seperations less than MAX_STARLANE_LENGTH
+    Then returns the clusters of stars when requested
+    """
+    def __init__(self, positions, adjacency_grid):
+        self.adjacency_grid = adjacency_grid
+        self.dsets = DisjointSets()
+
+        for pos in positions:
+            self._add(pos)
+
+        self.clusters = set()
+        for cluster in self.dsets.complete_sets():
+            # Decorate each cluster with its length to speed sorting
+            self.clusters.add((len(cluster), frozenset(cluster)))
+
+    def __len__(self):
+        return len(self.clusters)
+
+    def _add(self, pos):
+        """Adds pos."""
+        for neighbor in self.adjacency_grid.neighbors(pos):
+            # print "Clusterer adding ", pos,  " and ", neighbor
+            self.dsets.link(pos, neighbor)
+
+    def smallest_isolated_cluster(self):
+        """
+        Return None if all of the positions are in one cluster
+        or the smallest cluster
+        """
+        # Sort (min) and return the undecorated cluster.
+        return None if len(self.clusters) < 2 else min(self.clusters)[1]
+
+    def stitch_clusters(self, p1, p2, stitches):
+        """
+        Stitch the clusters containing ''p1'' and ''p2''
+        together with the positions in ''stitches''
+
+        This assumes that the stitching is correctly formed.
+        ''p1'' and ''p2'' are the closest positions between
+        the clusters and the positions in ''stitches'' are only
+        between ''p1'' and ''p2''
+        """
+        len_dset1 = None
+        len_dset2 = None
+        for len_dset in self.clusters:
+            if p1 in len_dset[1]:
+                len_dset1 = len_dset
+            if p2 in len_dset[1]:
+                len_dset2 = len_dset
+        assert len_dset1 and len_dset2, "p1 and p2 must be points in disjoint sets of positions"
```

Comment:
See my other comment a few lines down. No reason to fail hard here, just log an error and abort the insertion of additional systems to enforce the max starlane length limit. This will result in getting a mao with overly long starlanes, but the player can still play. If you fail with an assertion here, the entire universe generation process will be aborted, resulting in an unplayable map.